sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -315,18 +319,47 @@ def release(self, instruction_id):
     """
     descriptor_id, processor = self.active_bundle_processors.pop(instruction_id)
     processor.reset()
+    self.last_access_time[descriptor_id] = time.time()
     self.cached_bundle_processors[descriptor_id].append(processor)
 
   def shutdown(self):
     """
     Shutdown all ``BundleProcessor``s in the cache.
     """
+    if self.periodic_shutdown:
+      self.periodic_shutdown.cancel()
+      self.periodic_shutdown.join()
+      self.periodic_shutdown = None
+
     for instruction_id in self.active_bundle_processors:
       self.active_bundle_processors[instruction_id][1].shutdown()
       del self.active_bundle_processors[instruction_id]
     for cached_bundle_processors in self.cached_bundle_processors.values():
-      while len(cached_bundle_processors) > 0:
-        cached_bundle_processors.pop().shutdown()
+      BundleProcessorCache._shutdown_cached_bundle_processors(
+          cached_bundle_processors)
+
+  def _schedule_periodic_shutdown(self):
+    def shutdown_inactive_bundle_processors():
+      for descriptor_id, last_access_time in self.last_access_time.items():
+        if time.time() - last_access_time > 60:
+          BundleProcessorCache._shutdown_cached_bundle_processors(
+              self.cached_bundle_processors[descriptor_id])
```

Comment:
Don't we have to remove the bundle processor list from the dictionary? Otherwise we may access a cached shutdown bundle processor.

Reply:
The bundle processor will be popped from the bundle processor list before shutdown and the pop operation is thread safe and so I think this will not happen. What's your thought?

Reply:
True, since the pop on the list is thread-safe, we can't wind up with a cached bundle processor being in use.